radixdlt/src/main/java/com/radixdlt/consensus/safety/SafetyRules.java

```diff
@@ -20,47 +20,81 @@
 import com.google.common.annotations.VisibleForTesting;
 import com.google.inject.Inject;
 import com.google.inject.name.Named;
-import com.radixdlt.common.EUID;
 import com.radixdlt.consensus.QuorumCertificate;
 import com.radixdlt.consensus.Vertex;
 import com.radixdlt.consensus.VertexMetadata;
+import com.radixdlt.consensus.VertexStore;
+import com.radixdlt.consensus.View;
 import com.radixdlt.consensus.Vote;
-
+import com.radixdlt.crypto.CryptoException;
+import com.radixdlt.crypto.ECDSASignature;
+import com.radixdlt.crypto.ECKeyPair;
 import com.radixdlt.crypto.Hash;
+
 import java.util.Objects;
+import java.util.Optional;
 
 /**
  * Manages safety of the protocol.
- * TODO: Add storage of private key of node here
  */
 public final class SafetyRules {
-	private final EUID self;
+	private final ECKeyPair selfKey; // TODO remove signing/address to separate identity management
 
+	private final VertexStore vertexStore;
 	private SafetyState state;
 
 	@Inject
-	public SafetyRules(@Named("self") EUID self, SafetyState initialState) {
-		this.self = Objects.requireNonNull(self);
-		this.state = new SafetyState(initialState.lastVotedRound, initialState.lockedRound);
+	public SafetyRules(@Named("self") ECKeyPair selfKey,
+	                   VertexStore vertexStore,
+	                   SafetyState initialState) {
+		this.selfKey = Objects.requireNonNull(selfKey);
+		this.vertexStore = Objects.requireNonNull(vertexStore);
+		this.state = new SafetyState(initialState);
 	}
 
-	@VisibleForTesting
-	Hash getCommittedVertexId(Vertex vertex) {
-		if (vertex.getRound().equals(vertex.getQC().getRound().next())
-			&& vertex.getQC().getRound().equals(vertex.getQC().getParentRound().next())) {
-			return vertex.getQC().getVertexMetadata().getParentId();
-		}
-		return null;
+	/**
+	 * Process a vertex
+	 * @param vertex The vertex
+	 * @return the now-committed aid, if any
+	 */
+	public Optional<Hash> process(Vertex vertex) {
+		return process(vertex.getQC());
 	}
 
 	/**
-	 * Process a quorum certificate
+	 * Process a QC seen at a certain view
 	 * @param qc The quorum certificate
+	 * @return the now-committed aid, if any
 	 */
-	public void process(QuorumCertificate qc) {
-		if (qc.getParentRound().compareTo(this.state.lockedRound) > 0) {
-			this.state = this.state.withLockedRound(qc.getParentRound());
+	public Optional<Hash> process(QuorumCertificate qc) {
+		// pre-commit phase on vertex's parent if there is a newer 1-chain
+		// keep highest 1-chain as the current "generic" QC
+		if (qc.getView().compareTo(this.state.getGenericView().orElse(View.of(0L))) > 0) {
+			this.state = this.state.withGenericQC(qc);
 		}
+
+		// commit phase on vertex's grandparent if there is a newer 2-chain
+		// keep the highest 2-chain as the locked QC
+		Vertex parent = vertexStore.getVertex(qc.getVertexMetadata().getId());
+		if (parent != null && parent.getQC() != null)  { // genesis QC may be null
```

Comment:
parent != null should be an IllegalStateException
parent.getQC(), genesis QC should never go through process() so this should be IllegalStateException as well

Reply:
Implemented on latest